features/netsocket/emac-drivers/TARGET_STM_EMAC/TARGET_STM32F4/TARGET_UBLOX_EVK_ODIN_W2/wifi_emac/wifi_emac.cpp

```diff
@@ -0,0 +1,304 @@
+}
+
+uint8_t handleWlanTargetGetDataFrameTID(cbWLANTARGET_dataFrame* frame)
+{
+    (void)frame;
+    return (uint8_t)cbWLAN_AC_BE;
+}
+
+WIFI_EMAC::WIFI_EMAC()
+{
+    emac_link_input_cb = NULL;
+    emac_link_state_cb = NULL;
+    cbWLANTARGET_registerCallbacks((cbWLANTARGET_Callback*)&_wlanTargetCallback);
+}
+
+void send_wlan_packet(void *buf)
+{
+    cbWLAN_sendPacket(buf);
+}
+
+bool WIFI_EMAC::link_out(emac_mem_buf_t *buf)
+{
+    EMACMemoryManager *mem = WIFI_EMAC::get_instance().memory_manager;
+
+    // Break call chain to avoid the driver affecting stack usage for the IP stack thread too much
+    emac_mem_buf_t *new_buf = mem->alloc_pool(mem->get_total_len(buf), 0);
+    if (new_buf != NULL) {
+        mem->copy(new_buf, buf);
+        int id = cbMAIN_getEventQueue()->call(send_wlan_packet, new_buf);
+        if (id != 0) {
+            cbMAIN_dispatchEventQueue();
+        } else {
+            mem->free(new_buf);
+        }
+    }
+    mem->free(buf);
+    return true;
+}
+
+bool WIFI_EMAC::power_up()
+{
+    /* Initialize the hardware */
+    /* No-op at this stage */
+    return true;
+}
+
+uint32_t WIFI_EMAC::get_mtu_size() const
+{
+    return WIFI_EMAC_MTU_SIZE;
+}
+
+void WIFI_EMAC::get_ifname(char *name, uint8_t size) const
+{
+    memcpy(name, _ifname, (size < sizeof(_ifname)) ? size : sizeof(_ifname));
+}
+
+uint8_t WIFI_EMAC::get_hwaddr_size() const
+{
+    return sizeof(cbWLAN_MACAddress);
+}
+
+bool WIFI_EMAC::get_hwaddr(uint8_t *addr) const
+{
+    cbOTP_read(cbOTP_MAC_WLAN, sizeof(cbWLAN_MACAddress), addr);
+    return true;
+}
+
+void WIFI_EMAC::set_hwaddr(const uint8_t *addr)
+{
+  /* No-op at this stage */
+}
+
+void WIFI_EMAC::set_link_input_cb(emac_link_input_cb_t input_cb)
+{
+    emac_link_input_cb = input_cb;
+
+    cbMAIN_driverLock();
+    cbWLAN_registerPacketIndicationCallback(handleWlanPacketIndication, NULL);
+    cbMAIN_driverUnlock();
+}
+
+void WIFI_EMAC::set_link_state_cb(emac_link_state_change_cb_t state_cb)
+{
+    emac_link_state_cb = state_cb;
+
+    cbMAIN_driverLock();
+    cbWLAN_registerStatusCallback(handleWlanStatusIndication, NULL);
+    cbMAIN_driverUnlock();
+}
+
+void WIFI_EMAC::power_down()
+{
+  /* No-op at this stage */
+}
+
+void WIFI_EMAC::set_memory_manager(EMACMemoryManager &mem_mngr)
+{
+    memory_manager = &mem_mngr;
+}
+
+uint32_t WIFI_EMAC::get_align_preference() const
+{
+    return 1; // TODO not sure if there is a requirement but don't think so
+}
+
+void WIFI_EMAC::add_multicast_group(const uint8_t *address)
+{
+    // TODO anything to do here for WiFi?
+}
+
+void WIFI_EMAC::remove_multicast_group(const uint8_t *address)
+{
+    // TODO anything to do here for WiFi?
+}
+
+void WIFI_EMAC::set_all_multicast(bool all)
+{
+    // TODO anything to do here for WiFi?
+}
+
+WIFI_EMAC &WIFI_EMAC::get_instance() {
+    static WIFI_EMAC emac;
+    return emac;
+}
+
+// Weak so a module can override
+MBED_WEAK EMAC &EMAC::get_default_instance()
```

Comment:
You should not be defining yourself as the default EMAC instance. That's only useful for Ethernet-type interfaces that can be run without configuration - it lets `EthernetInterface` pick you up.

(And this is causing double-definition compile errors versus the definition in stm32xx_emac.cpp)

Reply:
Corrected EMAC tests to not to refer to "get_default_instance()" in this pull request: https://github.com/ARMmbed/mbed-os/pull/6941
Update test environment with that when removing the "get_default_instance()"